### DISCUSSION TITLE
Using yarp python bindings, how do you convert a yarp.image to an np.array?

### DISCUSSION THREAD

**Discussion by arrenglover** [robotology/community#544] [Q&A] [answered]
Using YARP in python, consider reading a yarp image over the network:

```python
import cv2
import yarp
myport = yarp.BufferedPortImageMono()
myimage = myport.read()
```
I would eventually like to visualise the data with opencv, but am interested in how to convert the data to numpy array. The following didn't work:

```python
cv2.imshow("image", myimage)
cv2.waitkey()
```

and I guess there is some function I can call to convert the data into a format known by numpy and/or opencv.

**Comment by pattacini** [comment]
Just mentioning @diegoferigo and @ddetommaso who, I know, are active Python developers.

**Comment by vtikha** [comment, marked as answer]
@arrenglover, maybe the [yarp image example](https://github.com/robotology/yarp/blob/master/bindings/python/examples/yarpImage.py) in the bindings section can help. It converts the image data to a numpy array.

**Reply by diegoferigo** [reply]
AFAIK this is the preferred way, and it is the only one not involving a copy when converting to numpy. However, using raw buffers in Python is quite advanced and error prone, use with care.

In other projects, like iDynTree, we provide [`fromPython|toNumPy`](https://github.com/robotology/idyntree/blob/master/bindings/python/python.i) methods for the conversion, but they always involve a copy, which can be suboptimal for big chunks of data as images. Zero copy and numpy support are two of the best feature IMO of pybind11, but we are relying on SWIG (which has other advantages) and I don't think this will change anytime soon.

Disclaimer: I don't work with images and never investigated too much on this. When I want to convert from YARP types to python, I typically convert by copy, since the data I need is never too big:


```python
def to_list(yarp_vector: yarp.DVector) -> List[float]:
    return [yarp_vector[i] for i in range(yarp_vector.size())]
```

**Reply by arrenglover** [reply]
great! thanks @vtikha and @diegoferigo. This is what I was looking for. I could only find examples that used a standard port and not the buffered port. It seems there is just one extra line to use the bufferedport. We are still having troubles with the setExternal() function of the yarp.Image but I think it must be something on our system, because the example you have linked seg faults at this point. We'll test further.

**Reply by Arya07** [reply]
Hi @arrenglover, the segfault that you are having with the setExternal() might be related to the python3 minor version you are using. Try to replace line 32 in the `yarpImage.py` with the following line:
```
in_buf_image.setExternal(in_buf_array.data, in_buf_array.shape[1], in_buf_array.shape[0])
```
and line 38 with:
```
 out_buf_image.setExternal(out_buf_array.data, out_buf_array.shape[1], out_buf_array.shape[0])
```
(i.e., adding `.data` to the original lines )

if that was the problem (which happened to be in my case), these changes should fix it. I hope it might help.

**Reply by arrenglover** [reply]
awesome thanks @Arya07, yes I suspect some version mismatch. We'll try this :+1: @gaurvigoyal

**Reply by gaurvigoyal** [reply]
Thanks @Arya07! Adding <<.data>> mitigated the segfault but the Assertion in line 47 still fails. 

`assert in_buf_array.__array_interface__['data'][0] == in_buf_image.getRawImage().__int__()`

The read image is not copied to the new address. Is there some other change also required for this to work?

Solution: The size of the image being received should match the size of the input numpy array precisely. Then the assertion is valid, and the array has the received image.
